packages/platforms/accton/x86-64/x86-64-accton-as5916-54xks/modules/builds/x86-64-accton-as5916-54xks-sfp.c

```diff
@@ -1342,6 +1369,111 @@ static ssize_t set_qsfp_lpmode(struct device *dev, struct device_attribute *da,
     return status;
 }
 
+static char *
+__strtok_r(char *s, const char *delim, char **last)
+{
+	char *spanp, *tok;
+	int c, sc;
+
+	if (s == NULL && (s = *last) == NULL)
+		return (NULL);
+
+	/*
+	 * Skip (span) leading delimiters (s += strspn(s, delim), sort of).
+	 */
+cont:
+	c = *s++;
+	for (spanp = (char *)delim; (sc = *spanp++) != 0;) {
+		if (c == sc)
+			goto cont;
+	}
+
+	if (c == 0) {		/* no non-delimiter characters */
+		*last = NULL;
+		return (NULL);
+	}
+	tok = s - 1;
+
+	/*
+	 * Scan token (scan for delimiters: s += strcspn(s, delim), sort of).
+	 * Note that delim must have one NUL; we stop if we see that, too.
+	 */
+	for (;;) {
+		c = *s++;
+		spanp = (char *)delim;
+		do {
+			if ((sc = *spanp++) == c) {
+				if (c == 0)
+					s = NULL;
+				else
+					s[-1] = '\0';
+				*last = s;
+				return (tok);
+			}
+		} while (sc != 0);
+	}
+	/* NOTREACHED */
+}
+
+/* Command Format:
+   <Port> <Slave Address> <Data Count> <Register 1> <Data 1 High> <Data 1 Low> <Register 2> <Data 2 High> <Data 2 Low> ...
+ */
+static ssize_t set_phy(struct device *dev, struct device_attribute *da,
+			const char *buf, size_t count)
+{
+    int i = 0;
+	int status;
+    char *token = NULL, *last = NULL, *delim = " ";
+    struct sfp_phy_write_data wdata;
+    char command[NUM_OF_PHY_REGISTERS * 3 + 3] = {0};
+    char *ipmi_phy_tx_data = (char *)&wdata;
+
+    memset(&wdata, 0, sizeof(wdata));
+    memcpy(command, buf, sizeof(command));
+
+    /* Parsing command into tokens */
+    token = __strtok_r(command, delim, &last);
+    while (token != NULL) {
+        long param = 0;
+
+    	status = kstrtol(token, 0, &param);
+    	if (status) {
+    		return status;
+    	} 
+
+        ipmi_phy_tx_data[i++] = param;
```

Comment:
should you not have a range check here, and return an error for an out of range parameter?

Reply:
Done for adding the range check.

```suggestion
        /* Validate each param length */
        if (param > 0xFF) {
            return -EINVAL;
        }
```

Reply:
Fix the command leak and range checking.
Please help verify.